drivers/sensor/ene_tach_kb1200/tach_ene_kb1200.c

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright (c) 2024 ENE Technology Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT ene_kb1200_tach
+
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/drivers/sensor.h>
+#include <reg/tacho.h>
+
+/* Device config */
+struct tach_kb1200_config {
+	/* tachometer controller base address */
+	uintptr_t base;
+	/* number of pulses (holes) per	round of tachometer's input (encoder) */
+	int pulses_per_round;
+	/* sampling clock timing of tachometer (us) */
+	int sample_clk;
+	const struct pinctrl_dev_config *pcfg;
+};
+
+/* Driver data */
+struct tach_kb1200_data {
+	/* Captured counts of tachometer */
+	uint32_t capture;
+};
+
+/* TACH	local functions	*/
+static int tach_kb1200_configure(const struct device *dev)
+{
+	const struct tach_kb1200_config *const config = dev->config;
+	TACHO_T *tacho = (TACHO_T *)config->base;
+	uint8_t sample_us = 0;
+
+	/* Configure clock module and its frequency of tachometer */
+	switch (config->sample_clk) {
+	case 2:
+		sample_us = TACHO_MONITOR_CLK_2US;
+		break;
+	case 8:
+		sample_us = TACHO_MONITOR_CLK_8US;
+		break;
+	case 16:
+		sample_us = TACHO_MONITOR_CLK_16US;
+		break;
+	case 64:
+		sample_us = TACHO_MONITOR_CLK_64US;
+		break;
+	default:
+		return -ENOSYS;
+	}
+	tacho->TACHOCFG = (sample_us << 4) | TACHO_FUNCTION_ENABLE;
+
+	return 0;
+}
+
+/* TACH	api functions */
+int tach_kb1200_sample_fetch(const struct device *dev, enum sensor_channel chan)
+{
+	ARG_UNUSED(chan);
+	struct tach_kb1200_data *const data = dev->data;
+	const struct tach_kb1200_config *const config = dev->config;
+	TACHO_T *tacho = (TACHO_T *)config->base;
```

Comment:
note: all these casts would not be needed if you just have `TACHO_T *` in config (applies to other drivers)